Geofence Actions

dagar opened the pull request:
Here's a basic version of geofence violation -> RTL. I'm opening this PR now to discuss the desired behaviour.

This satisfies the basic functionality I want, but what does everyone else want or need? So far this just kicks you into RTL once you violate the geofence. You can escape by changing any switch.

**Ideas/Questions**
 - should we first warn and then only take control if still outside?
 - param to select either warning or RTL
 - param to choose between RTL or LOITER
 - ability to retake control without changing mode switches (like touching the sticks)
 - ability to retake control while still outside of the geofence
    - would the geofence then kick in again if you're still outside after a given time period?
 - more sophisticated fixed wing behaviour (similar to APM)
 - qgroundcontrol gui support?

Thoughts?

dagar commented:
I'll keep going with this if no one has a strong opinion about functionality.
Next I'll replace GF_MODE with GF_ACTION

**GF_ACTION**
 - 0 - Disable
 - 1 - Warning (default)
 - 2 - Loiter
 - 3 - RTL

I don't think we need to explicitly set the modes because if geofence.txt isn't present or GF_MAX_HOR_DIST/GF_MAX_VER_DIST aren't set they're just ignored.

LorenzMeier commented:
Hey Daniel! This is super cool! I was just swamped totally with travel to IROS. One comment: What would you associate in an english text message with "gf action"? I found it a great param name, made my day 8).

LorenzMeier commented on src/modules/commander/commander.cpp:
```diff
@@ -2408,7 +2412,18 @@ set_main_state_rc(struct vehicle_status_s *status_local, struct manual_control_s
 
-	/* if offboard is set allready by a mavlink command, abort */
+	/* check if mode switches have changed */
+	if ((_prev_sp_man.mode_switch == sp_man->mode_switch)
```
I recall having this implemented also on master, need to check where.

dagar replied:
I'll rebase.

dagar commented:
This needs thorough testing, but it implements the following.

**GF_ACTION**

0 - Disable
1 - Warning (default, and current behaviour)
2 - Loiter
3 - RTL
4 - flight termination

- you can retake control after geofence violation by flipping the mode switches OR if you were in a manual/assist mode before violation and you move the sticks
- if you're outside the fence and retake control the geofence will retrigger after 10 seconds

LorenzMeier commented:
Thanks! @tumbili Could you add that to your test queue in SITL and later outdoors?

dagar commented:
I tested this today and it mostly works, but retaking control with the sticks is way too sensitive.

dagar commented:
I rebased on master and changed the escape from RTL/LOITER logic to require the sticks to increase. The idea being that if you pass the boundary you should be able to let the sticks fall back to neutral without reverting the failsafe.
I also refactored the logic to make it a bit easier to follow and tested in HIL.

LorenzMeier commented:
Hmm. I think that needs some thought - if I was flying with max pitch (for a quad) and high throttle out of the fence as in a FPV racer, how do I disengage the fence then?

LorenzMeier commented:
The original headers in there had a bad naming scheme, sorry for the confusion. The official MAVLink frame is NED and all of QGCs local coordinate logic is NED.

dagar commented:
@LorenzMeier how would you want to disengage in that situation? You still get out by wiggling the sticks  (down then up a bit), but this way it doesn't revert if I let them fall to neutral.
The mode switch is always respected.

LorenzMeier commented:
True and fair enough. Let's get this in!

LorenzMeier commented:
Cherry picked to master, thanks!
